File: Office-Hours/Excel-Python/simple_excel_writer.py
from openpyxl import Workbook

from examples.roster import Roster


def demo_simple_table(destination: str = "report.xlsx") -> None:
    """
    Output a 3x3 table with the cells labeled as A1, A2... C2, C3
    """
    workbook = Workbook()
    sheet = workbook.create_sheet(title="Report")

    # Output 3 Rows (A, B, C)
    for idx_row in range(1, 3):
        for idx_col in range(ord("A"), ord("D")):
            # Set up the cell data
            cell_name = f"{chr(idx_col)}{idx_row}"

            # Create the cell and write to it
            sheet[cell_name] = cell_name

    # Write and save the Excel document
    workbook.save(destination)


def demo_roster_table(roster: Roster, destination: str = "report.xlsx") -> None:
    """
    Output a table with student names in the first column and grades in the
    second column.

    TODO: replace hardcoded "4.0" grades with actual grade values.
    """

    workbook = Workbook()
    sheet = workbook.create_sheet(title="Report")

    # Output the Header
    sheet.append(["Student", "Grade"])

    # Output all students
    for stu in roster.list_enrolled_students():
        # Create and output a student row
        row = [stu.name, 4.0]
        sheet.append(row)

    # Write and save the Excel document
    workbook.save(destination)

File: Office-Hours/Excel-Rust/src/simple_excel_writer.rs
use rust_xlsxwriter::*;

use crate::prelude::{Roster, Student};

///
/// Output a 3x3 table with the cells labeled as A1, A2... C2, C3
///
pub fn demo_simple_table(destination: &str) -> eyre::Result<()> {
    let mut workbook = Workbook::new();
    let sheet = workbook.add_worksheet().set_name("Report")?;

    // Output 3 Rows (A, B, C)
    for idx_row in 0..2 {
        for (idx_col, letter_col) in (0..3).zip(['A', 'B', 'C']) {
            let cell_entry = format!("{}{}", letter_col, idx_row + 1);
            sheet.write(idx_row, idx_col, cell_entry)?;
        }
    }

    // Write and save the Excel document
    workbook.save(destination)?;

    Ok(())
}

/// Output a table with student names in the first column and grades in the
/// second column.
///
/// TODO: replace hardcoded "4.0" grades with actual grade values.
///
pub fn demo_roster_table(roster: Roster, destination: &str) -> eyre::Result<()> {
    let mut workbook = Workbook::new();
    let sheet = workbook.add_worksheet().set_name("Report")?;

    // Output the Header
    sheet.write(0, 0, "Student")?;
    sheet.write(0, 1, "Grade")?;

    // Output all students
    for (idx_row, stu) in (1..).zip(roster.iter()) {
        // Create and output a student row
        sheet.write(idx_row.try_into()?, 0, &stu.name)?;
        sheet.write_number(idx_row.try_into()?, 1, 4.0)?;
    }

    // Write and save the Excel document
    workbook.save(destination)?;

    Ok(())
}

